File: packages/nativescript-masonkit/src-native/mason-android/masonkit/src/main/java/org/nativescript/mason/masonkit/MinMax.kt
package org.nativescript.mason.masonkit;

sealed class MinMax(
  val min: MinSizing,
  val max: MaxSizing
) {
  data class Points(val points: Float) :
    MinMax(MinSizing.Points(points), MaxSizing.Points(points))

  data class Percent(var percentage: Float) :
    MinMax(MinSizing.Percent(percentage), MaxSizing.Percent(percentage))

  data class Fraction(var fraction: Float) :
    MinMax(MinSizing.Auto, MaxSizing.Fraction(fraction))

  data class FitContent(var points: Float) :
    MinMax(MinSizing.Auto, MaxSizing.FitContent(points))

  data class FitContentPercent(var percent: Float) :
    MinMax(MinSizing.Auto, MaxSizing.FitContentPercent(percent))

  data object Auto : MinMax(MinSizing.Auto, MaxSizing.Auto)
  data object MinContent : MinMax(MinSizing.MinContent, MaxSizing.MinContent)
  data object MaxContent : MinMax(MinSizing.MaxContent, MaxSizing.MaxContent)

  data class Values(
    val minVal: MinSizing,
    val maxVal: MaxSizing
  ) : MinMax(minVal, maxVal)

  companion object {
    @JvmStatic
    fun fromTypeValue(minType: Int, minValue: Float, maxType: Int, maxValue: Float): MinMax? {
      val min = when (minType) {
        0 -> MinSizing.Auto
        1 -> MinSizing.MinContent
        2 -> MinSizing.MaxContent
        3 -> MinSizing.Points(minValue)
        4 -> MinSizing.Percent(minValue)
        else -> null
      }

      val max = when (maxType) {
        0 -> MaxSizing.Auto
        1 -> MaxSizing.MinContent
        2 -> MaxSizing.MaxContent
        3 -> MaxSizing.Points(maxValue)
        4 -> MaxSizing.Percent(maxValue)
        5 -> MaxSizing.Fraction(maxValue)
        6 -> MaxSizing.FitContent(maxValue)
        7 -> MaxSizing.FitContentPercent(maxValue)
        else -> null
      }

      return if (min != null && max != null) Values(min, max) else null
    }
  }

  val minType: Int get() = min.type
  val minValue: Float get() = min.value
  val maxType: Int get() = max.type
  val maxValue: Float get() = max.value

  val cssValue: String
    get() = when (this) {
      Auto -> "auto"
      is Fraction -> "${maxValue}fr"
      MaxContent -> "max-content"
      MinContent -> "min-content"
      is Percent -> "${"%.2f".format(minValue * 100)}%"
      is Points -> "${minValue}px"
      is FitContent -> "fit-content(${maxValue}px)"
      is FitContentPercent -> "fit-content(${"%.2f".format(maxValue * 100)}%)"
      is Values -> when {
        minVal == MinSizing.Auto && maxVal == MaxSizing.Auto -> "auto"
        minVal == MinSizing.MinContent && maxVal == MaxSizing.MinContent -> "min-content"
        minVal == MinSizing.MaxContent && maxVal == MaxSizing.MaxContent -> "max-content"
        minVal == MinSizing.Auto && maxVal is MaxSizing.Fraction -> "${maxValue}fr"
        minVal == MinSizing.Auto && maxVal is MaxSizing.FitContent -> "fit-content(${maxValue}px)"
        minVal == MinSizing.Auto && maxVal is MaxSizing.FitContentPercent ->
          "fit-content(${"%.2f".format(maxValue * 100)}%)"

        minType == maxType && minValue == maxValue -> when (minVal) {
          is MinSizing.Percent -> "${"%.2f".format(minVal.percentage * 100)}%"
          is MinSizing.Points -> "${minVal.points}px"
          else -> ""
        }

        else -> {
          val minStr = when (minVal) {
            MinSizing.Auto -> "auto"
            MinSizing.MaxContent -> "max-content"
            MinSizing.MinContent -> "min-content"
            is MinSizing.Percent -> "${"%.2f".format(minVal.percentage * 100)}%"
            is MinSizing.Points -> "${minVal.points}px"
          }

          val maxStr = when (maxVal) {
            MaxSizing.Auto -> "auto"
            MaxSizing.MaxContent -> "max-content"
            MaxSizing.MinContent -> "min-content"
            is MaxSizing.Percent -> "${"%.2f".format(maxVal.percentage * 100)}%"
            is MaxSizing.Points -> "${maxVal.points}px"
            is MaxSizing.Fraction -> "${maxValue}fr"
            // CSS doesn’t allow fit-content inside minmax, use fallback
            is MaxSizing.FitContent, is MaxSizing.FitContentPercent -> "auto"
          }

          "minmax($minStr, $maxStr)"
        }
      }
    }
}

val Array<MinMax>.jsonValue: String
  get() = Mason.gson.toJson(this)

val Array<MinMax>.cssValue: String
  get() = joinToString(" ") { it.cssValue }


typealias TrackSizingFunction = MinMax

File: packages/nativescript-masonkit/src-native/mason-ios/Mason/Mason/MinMax.swift
//
//  MinMax.swift
//  Mason
//
//  Created by Osei Fortune on 01/01/2023.
//

import Foundation

@objcMembers
@objc(MinMax)
public class MinMax: NSObject, Codable {
    let min: MinSizing
    let max: MaxSizing
    
    public init(_ min: MinSizing, _ max: MaxSizing) {
        self.min = min
        self.max = max
    }
    
    public static func Points(points: Float) -> MinMax {
        return MinMax(.Points(points), .Points(points))
    }
    
    public static func Percent(percent: Float) -> MinMax {
        return MinMax(.Percent(percent), .Percent(percent))
    }
    
    public static func Flex(flex: Float) -> MinMax {
        return MinMax(.Auto, .Fraction(flex))
    }
    
    public static func FitContent(fit: Float) -> MinMax {
        return MinMax(.Auto, .FitContent(fit))
    }
    
    public static func FitContentPercent(fit: Float) -> MinMax {
        return MinMax(.Auto, .FitContentPercent(fit))
    }
    
    public static let Auto = MinMax(.Auto, .Auto)
    
    public static func fromTypeValue(_ minType: Int, _ minValue: Float, _ maxType: Int, _ maxValue: Float) -> MinMax? {
        var min: MinSizing? = nil
        
        switch (minType) {
        case 0: min = MinSizing.Auto
            break
        case 1: min = MinSizing.MinContent
            break
        case 2: min = MinSizing.MaxContent
            break
        case 3: min = MinSizing.Points(minValue)
            break
        case 4: min = MinSizing.Percent(minValue)
            break
        default:
            min = nil
        }
        
        var max: MaxSizing? = nil
        
        switch (maxType) {
        case 0: max = MaxSizing.Auto
            break
        case 1: max = MaxSizing.MinContent
            break
        case 2: max = MaxSizing.MaxContent
            break
        case 3: max = MaxSizing.Points(maxValue)
            break
        case 4: max = MaxSizing.Percent(maxValue)
            break
        case 5 : max = MaxSizing.Fraction(maxValue)
            break
        case 6 : max = MaxSizing.FitContent(maxValue)
            break
        case 7 : max =  MaxSizing.FitContentPercent(maxValue)
            break
        default:
            max = nil
        }
        
        guard let min = min else{return nil}
        guard let max = max else{return nil}
        
        
        return MinMax(min, max)
    }
    
    
    
    var minType: Int32{
        get {
            return min.type
        }
    }
    
    var minValue: Float{
        get {
            return min.value
        }
    }
    
    
    var maxType: Int32{
        get {
            return max.type
        }
    }
    
    
    var maxValue: Float{
        get {
            return max.value
        }
    }
    
    
    
    public var cssValue: String {
        get {
            switch((self.min.internalType, self.max.internalType)){
            case (.Auto, .Auto):
                return "auto"
            case (.MinContent, .MinContent):
                return "min-content"
            case (.MaxContent, .MaxContent):
                return "max-content"
            case (.Points, .Points):
                if(self.min.value == self.max.value){
                    return "\(self.min.value)px"
                }else {
                    return "minmax(\(self.min.value)px, \(self.max.value))px"
                }
            case (.Percent, .Percent):
                if(self.min.value == self.max.value){
                    return "\(self.min.value)%"
                }else {
                    return "minmax(\(self.min.value)%, \(self.max.value))%"
                }
            case (.Auto, .Fraction):
                return "\(self.max.value)fr"
            case (.Auto, .FitContent):
                return "fit-content(\(self.max.value)px)"
            case (.Auto, .FitContentPercent):
                return "fit-content(\(self.max.value)%)"
            default:
                return "minmax(\(self.min.cssValue), \(self.max.cssValue))"
            }
        }
    }
    
    public var jsonValue: String? {
        do {
            return try String(data: encoder.encode(self), encoding: .utf8)
        }catch {return nil}
    }
    
    static func decodeMinValue(value: String) -> MinSizing? {
        switch(value){
        case "auto":
            return .Auto
        case "min-content":
            return .MinContent
        case "max-content":
            return .MaxContent
        default:
            if(value.contains("%")){
                return .Percent(Float(value.replacingOccurrences(of: "%", with: ""))!)
            }else if(value.contains("px")){
                return .Points(Float(value.replacingOccurrences(of: "px", with: ""))!)
            }
            
            return nil
            
        }
    }
    
    static func decodeMaxValue(value: String) -> MaxSizing? {
        switch(value){
        case "auto":
            return .Auto
        case "min-content":
            return .MinContent
        case "max-content":
            return .MaxContent
        default:
            if(value.starts(with: "fit-content(")){
                let fitContent = value
                    .replacingOccurrences(of: "fit-content(", with: "")
                    .replacingOccurrences(of: ")", with: "")
                    .trimmingCharacters(in: .whitespaces)
                
                if(value.contains("%")){
                    return .FitContentPercent(Float(fitContent.replacingOccurrences(of: "%", with: ""))!)
                }else if(value.contains("px")){
                    return .FitContent(Float(fitContent.replacingOccurrences(of: "px", with: ""))!)
                }else {
                    return nil
                }
            }else if(value.contains("fr")){
                return .Fraction(Float(value.replacingOccurrences(of: "fr", with: ""))!)
            }else if(value.contains("%")){
                return .Percent(Float(value.replacingOccurrences(of: "%", with: ""))!)
            }else if(value.contains("px")){
                return .Points(Float(value.replacingOccurrences(of: "px", with: ""))!)
            }
            
            return nil
            
        }
    }
    
    
    required public init(from decoder: Decoder) throws {
        min = .Auto
        max = .Auto
        let container = try decoder.singleValueContainer()
        let value = try container.decode(String.self)
        switch(value){
        case "auto":
            self.min.internalType = .Auto
            self.max.internalType = .Auto
            break
        case "min-content":
            self.min.internalType = .MinContent
            self.max.internalType = .MinContent
            break
        case "max-content":
            self.min.internalType = .MaxContent
            self.max.internalType = .MaxContent
            break
        default:
            if(value.starts(with: "fit-content(")){
                let fitContent = value
                    .replacingOccurrences(of: "fit-content(", with: "")
                    .replacingOccurrences(of: ")", with: "")
                    .trimmingCharacters(in: .whitespaces)
                
                if(value.contains("%")){
                    self.min.internalType = .Auto
                    self.max.internalType = .FitContentPercent
                    self.max.value = Float(fitContent.replacingOccurrences(of: "%", with: ""))!
                }else if(value.contains("px")){
                    self.min.internalType = .Auto
                    self.max.internalType = .FitContent
                    self.max.value = Float(fitContent.replacingOccurrences(of: "px", with: ""))!
                }else {
                    // todo throw
                }
            }else if(value.contains("fr")){
                self.min.internalType = .Auto
                self.max.internalType = .Fraction
                self.max.value =  Float(value.replacingOccurrences(of: "fr", with: ""))!
            }else if(value.starts(with: "minmax(")) {
                let split = value
                    .replacingOccurrences(of: "minmax(", with: "")
                    .replacingOccurrences(of: ")", with: "")
                    .trimmingCharacters(in: .whitespaces)
                    .split(separator: ",")
                
                let min = MinMax.decodeMinValue(value: String(split.first!))!
                
                let max = MinMax.decodeMaxValue(value: String(split.last!))!
                
                self.min.internalType = min.internalType
                self.min.value = min.value
                
                self.max.internalType = max.internalType
                self.max.value = max.value
                
                
            }else if(value.contains("%")){
                
                let value = Float(value.replacingOccurrences(of: "%", with: ""))!
                
                self.min.internalType = .Percent
                self.min.value = value
                self.max.internalType = .Percent
                self.max.value = value
            }else if(value.contains("px")){
                
                let value = Float(value.replacingOccurrences(of: "px", with: ""))!
                
                self.min.internalType = .Points
                self.min.value = value
                self.max.internalType = .Points
                self.max.value = value
                
            }
            
            throw NSError(domain: "Invalid type", code: 1000)
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(cssValue)
    }
}
